app/lib/dns.server.ts

```diff
@@ -0,0 +1,192 @@
+import {
+  Route53Client,
+  CreateHostedZoneCommand,
+  ChangeResourceRecordSetsCommand,
+  GetChangeCommand,
+} from "@aws-sdk/client-route-53";
+import { CreateHostedZoneResponse } from "@aws-sdk/client-route-53";
+import { isIPv4, isIPv6 } from "is-ip";
+
+export const route53Client = new Route53Client({
+  endpoint: process.env.AWS_ENDPOINT_URL || "http://localhost:5053",
+  region: process.env.AWS_REGION || "us-east-1",
+  credentials: {
+    accessKeyId: process.env.AWS_ACCESS_KEY_ID || "",
+    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY || "",
+    sessionToken: process.env.AWS_SESSION_TOKEN,
+  },
+});
+
+export const createHostedZone = async (domain: string) => {
+  try {
+    const command = new CreateHostedZoneCommand({
+      Name: domain,
+      CallerReference: new Date().toString(),
+    });
+    const response: CreateHostedZoneResponse = await route53Client.send(
+      command
+    );
+
+    if (!response.HostedZone?.Id) {
+      throw new Error("Missing hosted zone ID in AWS response");
+    }
+    return response.HostedZone?.Id.replace("/hostedzone/", "");
+  } catch (error) {
+    if (error instanceof Error) {
+      console.warn(error);
+      throw new Error(`Error while creating hosted zone, ${error.message}`);
+    }
+  }
+};
+
+export const createRecord = (
+  hostedZoneId: string,
+  type: RecordType,
+  name: string,
+  value: string
+) => {
+  try {
+    return upsertRecord(hostedZoneId, type, name, value);
+  } catch (error) {
+    if (error instanceof Error) {
+      throw new Error(
+        `Error occurred while creating resource record: ${error.message}`
+      );
+    }
+  }
+};
+
+export const upsertRecord = async (
+  hostedZoneId: string,
+  type: RecordType,
+  name: string,
+  value: string
+): Promise<string | undefined> => {
```

Comment:
You haven't followed the same pattern I showed you above with `CreateHostedZone` (i.e., how you do the error typing, using the specific type from AWS).  Can you please fix all of your methods in here to use the same pattern I showed you above?